tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -741,6 +755,691 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+{
+	TsFdwRelInfo *fpinfo;
+	TsFdwRelInfo *fpinfo_o;
+	TsFdwRelInfo *fpinfo_i;
+	ListCell *lc;
+	List *joinclauses;
+
+	/*
+	 * If either of the joining relations is marked as unsafe to pushdown, the
+	 * join can not be pushed down.
+	 */
+	fpinfo = fdw_relinfo_get(joinrel);
+	fpinfo_o = fdw_relinfo_get(outerrel);
+	fpinfo_i = fdw_relinfo_get(innerrel);
+
+	/* shouldn't happen in the current implementation */
+	if (!fpinfo_o || !fpinfo_o->pushdown_safe || !fpinfo_i || !fpinfo_i->pushdown_safe)
+		return false;
+
+	/*
+	 * If joining relations have local conditions, those conditions are
+	 * required to be applied before joining the relations. Hence the join can
+	 * not be pushed down (shouldn't happen in the current implementation).
+	 */
+	if (fpinfo_o->local_conds || fpinfo_i->local_conds)
+		return false;
+
+	fpinfo->server = fpinfo_o->server;
+
+	/*
+	 * Separate restrict list into join quals and pushed-down (other) quals.
+	 *
+	 * Join quals belonging to an outer join must all be shippable, else we
+	 * cannot execute the join remotely.  Add such quals to 'joinclauses'.
+	 *
+	 * Add other quals to fpinfo->remote_conds if they are shippable, else to
+	 * fpinfo->local_conds.  In an inner join it's okay to execute conditions
+	 * either locally or remotely; the same is true for pushed-down conditions
+	 * at an outer join.
+	 *
+	 * Note we might return failure after having already scribbled on
+	 * fpinfo->remote_conds and fpinfo->local_conds.  That's okay because we
+	 * won't consult those lists again if we deem the join unshippable.
+	 */
+	joinclauses = NIL;
+	foreach (lc, extra->restrictlist)
+	{
+		RestrictInfo *rinfo = lfirst_node(RestrictInfo, lc);
+		bool is_remote_clause = ts_is_foreign_expr(root, joinrel, rinfo->clause);
+
+		if (IS_OUTER_JOIN(jointype) && !RINFO_IS_PUSHED_DOWN(rinfo, joinrel->relids))
+		{
+			if (!is_remote_clause)
+				return false;
+			joinclauses = lappend(joinclauses, rinfo);
+		}
+		else
+		{
+			if (is_remote_clause)
+				fpinfo->remote_conds = lappend(fpinfo->remote_conds, rinfo);
+			else
+				fpinfo->local_conds = lappend(fpinfo->local_conds, rinfo);
+		}
+	}
+
+	if (fpinfo->local_conds != NIL)
+		return false;
+
+	/* Save the join clauses, for later use. */
+	fpinfo->joinclauses = joinclauses;
+
+	/*
+	 * deparseExplicitTargetList() isn't smart enough to handle anything other
+	 * than a Var.  In particular, if there's some PlaceHolderVar that would
+	 * need to be evaluated within this join tree (because there's an upper
+	 * reference to a quantity that may go to NULL as a result of an outer
+	 * join), then we can't try to push the join down because we'll fail when
+	 * we get to deparseExplicitTargetList().  However, a PlaceHolderVar that
+	 * needs to be evaluated *at the top* of this join tree is OK, because we
+	 * can do that locally after fetching the results from the remote side.
+	 */
+	foreach (lc, root->placeholder_list)
```

Comment:
PlaceHolderVars always make my head hurt... I see that this is also postgres_fdw code, but let's add a test for it. Codecov complains. We should think twice whether we need outer joins, might be easier to delete all this.

Reply:
I could not trigger this code in the tests because we don't support outer joins or joins on views with constants at the moment. We could remove this code for the current implementation. As I mentioned [here](https://github.com/timescale/timescaledb/pull/5212#discussion_r1095802858), if we decide to remove this kind of code, supporting these features later takes more effort because you need to compare our implementation and the PG upstream implementation line-by-line.

Reply:
OK, let's keep it for the ease of comparison. Maybe makes sense to add a comment and/or assert about this somewhere.

Reply:
As suggested, I added a comment that this code is currently not used.